### PR TITLE
Misc fixes and improvements

### PR DESCRIPTION
**Pull request by tdesveaux**
Some random smaller patches I did while working on other stuff.

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `newsfragments` directory (and read the `README.txt` in that directory)
* [n/a] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/process/workerforbuilder.py [hunk 1]
```diff
@@ -221,7 +222,8 @@ def attached(self, worker, commands):
         return super().attached(worker, commands)
 
     def substantiate(self, build):
-        return self.worker.substantiate(self, build)
+        if self.worker:
+            return self.worker.substantiate(self, build)
```
Comment (p12tic) [inline master/buildbot/process/workerforbuilder.py:226]: I think we should keep appropriate return when the condition is not matched, probably `defer.succeed(None)`

File: master/buildbot/process/build.py [hunk 2]
```diff
@@ -577,37 +578,26 @@ def addStepsAfterLastStep(self, step_factories):
         # Add the new steps to the end.
         self.steps.extend(self.setupBuildSteps(step_factories))
 
-    def getNextStep(self):
+    def getNextStep(self) -> IBuildStep | None:
         """This method is called to obtain the next BuildStep for this build.
-        When it returns None (or raises a StopIteration exception), the build
+        When it returns None, the build
         is complete."""
```
Comment (p12tic) [inline master/buildbot/process/build.py:585]: This line could be merged in the previous line

File: master/buildbot/process/build.py [hunk 3]
```diff
@@ -577,37 +578,26 @@ def addStepsAfterLastStep(self, step_factories):
         # Add the new steps to the end.
         self.steps.extend(self.setupBuildSteps(step_factories))
 
-    def getNextStep(self):
+    def getNextStep(self) -> IBuildStep | None:
         """This method is called to obtain the next BuildStep for this build.
-        When it returns None (or raises a StopIteration exception), the build
```
Comment (p12tic) [inline master/buildbot/process/build.py:582]: I think it makes sense to add to commit description that the behavior of handling StopIteration exception is not documented, thus it's fine to consider this internal API and change it without deprecation.

**Review by p12tic — CHANGES_REQUESTED** [review]
LGTM, but I had several easy to address comments.

File: master/buildbot/process/workerforbuilder.py (hunk 1 above)
Comment (tdesveaux) [reply master/buildbot/process/workerforbuilder.py:226]: Good catch.
Used `defer.succeed(False)` since `worker.substantiate` returns a `Deferred[bool]`.

**Review by p12tic — APPROVED** [review]

**Pull request merged**